source/extensions/filters/http/oauth/filter.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/extensions/filters/http/oauth/v3/oauth.pb.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/stats_macros.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/assert.h"
+#include "common/config/datasource.h"
+#include "common/http/rest_api_fetcher.h"
+
+#include "extensions/filters/http/common/pass_through_filter.h"
+#include "extensions/filters/http/oauth/oauth.h"
+#include "extensions/filters/http/oauth/oauth_client.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Oauth {
+
+class OAuth2Client;
+
+// Helper class used to fetch secrets (usually from SDS).
+class SecretReader {
+public:
+  virtual ~SecretReader() = default;
+  virtual std::string clientSecret() const PURE;
+  virtual std::string tokenSecret() const PURE;
+};
+
+class SDSSecretReader : public SecretReader {
+public:
+  SDSSecretReader(Secret::GenericSecretConfigProviderSharedPtr client_secret_provider,
+                  Secret::GenericSecretConfigProviderSharedPtr token_secret_provider, Api::Api& api)
+      : client_secret_provider_(client_secret_provider),
+        token_secret_provider_(token_secret_provider), api_(api) {}
+
+  std::string clientSecret() const override {
+    if (!client_secret_provider_) {
+      return "";
+    }
+    const auto* secret = client_secret_provider_->secret();
+    if (!secret) {
+      return "";
+    }
+    return Config::DataSource::read(secret->secret(), true, api_);
+  }
+
+  std::string tokenSecret() const override {
+    if (!token_secret_provider_) {
+      return "";
+    }
+    const auto* secret = token_secret_provider_->secret();
+    if (!secret) {
+      return "";
+    }
+    return Config::DataSource::read(secret->secret(), true, api_);
+  }
+
+private:
+  Secret::GenericSecretConfigProviderSharedPtr client_secret_provider_;
+  Secret::GenericSecretConfigProviderSharedPtr token_secret_provider_;
+  Api::Api& api_;
+};
+
+/**
+ * All stats for the OAuth filter. @see stats_macros.h
+ */
+// clang-format off
+#define ALL_OAUTH_FILTER_STATS(COUNTER)                                                         \
+  COUNTER(oauth_failure)                                                                        \
+  COUNTER(oauth_success)
```

Comment:
Let's also add
- `unauthenticated_rq` counter of how many unauthenticated requests (which we 302) that we receive
- `authentication_rejected` counter of how many requests presented credentials that we reject.